File: node-backend/app/services/identity_service.js
const { createPublicClient, http } = require('viem');
const { celo } = require('viem/chains');
const { IdentitySDK } = require('@goodsdks/citizen-sdk');
const config = require('../config');

class IdentityService {
  constructor() {
    this.publicClient = createPublicClient({
      chain: celo,
      transport: http(config.celoRpcUrl),
    });
    this.identitySDK = new IdentitySDK(this.publicClient, null, config.veragigEnv);
  }

  async getWhitelistedRoot(account) {
    try {
      const { isWhitelisted, root } = await this.identitySDK.getWhitelistedRoot(account);
      // We also need lastAuthenticated for the response
      // IdentitySDK might not expose it directly in getWhitelistedRoot, let's check if we can get it from the contract
      // The Python code calls the contract directly
      const lastAuth = await this.publicClient.readContract({
        address: config.identityContract,
        abi: [
          {
            "inputs": [{"internalType": "address", "name": "account", "type": "address"}],
            "name": "lastAuthenticated",
            "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
            "stateMutability": "view",
            "type": "function",
          }
        ],
        functionName: 'lastAuthenticated',
        args: [account],
      });

      return {
        is_whitelisted: isWhitelisted,
        root: root,
        last_authenticated_timestamp: Number(lastAuth),
      };
    } catch (error) {
      console.error('Error in getWhitelistedRoot:', error);
      return { is_whitelisted: false, root: null, last_authenticated_timestamp: 0 };
    }
  }

  async getExpiryData(account) {
    try {
      const lastAuth = await this.publicClient.readContract({
        address: config.identityContract,
        abi: [
          {
            "inputs": [{"internalType": "address", "name": "account", "type": "address"}],
            "name": "lastAuthenticated",
            "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
            "stateMutability": "view",
            "type": "function",
          },
          {
            "inputs": [],
            "name": "authenticationPeriod",
            "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
            "stateMutability": "view",
            "type": "function",
          }
        ],
        functionName: 'lastAuthenticated',
        args: [account],
      });

      const authPeriod = await this.publicClient.readContract({
        address: config.identityContract,
        abi: [
          {
            "inputs": [],
            "name": "authenticationPeriod",
            "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
            "stateMutability": "view",
            "type": "function",
          }
        ],
        functionName: 'authenticationPeriod',
      });

      const expiresAt = Number(lastAuth) + (Number(authPeriod) * 86400);
      const isExpired = Date.now() / 1000 > expiresAt;

      return {
        last_authenticated: Number(lastAuth),
        authentication_period_days: Number(authPeriod),
        expires_at: expiresAt,
        is_expired: isExpired,
      };
    } catch (error) {
      console.error('Error in getExpiryData:', error);
      return { last_authenticated: 0, authentication_period_days: 0, expires_at: 0, is_expired: true };
    }
  }
}

let instance = null;
const getIdentityService = () => {
  if (!instance) {
    instance = new IdentityService();
  }
  return instance;
};

module.exports = { getIdentityService, IdentityService };

File: backend/app/services/identity_service.py
from web3 import Web3
from app.config import get_settings

settings = get_settings()

IDENTITY_ABI = [
    {
        "inputs": [{"internalType": "address", "name": "account", "type": "address"}],
        "name": "getWhitelistedRoot",
        "outputs": [
            {"internalType": "address", "name": "root", "type": "address"},
            {"internalType": "bool", "name": "isWhitelisted", "type": "bool"},
        ],
        "stateMutability": "view",
        "type": "function",
    },
    {
        "inputs": [{"internalType": "address", "name": "account", "type": "address"}],
        "name": "lastAuthenticated",
        "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
        "stateMutability": "view",
        "type": "function",
    },
    {
        "inputs": [],
        "name": "authenticationPeriod",
        "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
        "stateMutability": "view",
        "type": "function",
    },
]


class IdentityService:
    def __init__(self):
        self.w3 = Web3(Web3.HTTPProvider(settings.celo_rpc_url))
        self.contract = self.w3.eth.contract(
            address=Web3.toChecksumAddress(settings.identity_contract),
            abi=IDENTITY_ABI,
        )

    async def get_whitelisted_root(self, account: str) -> dict:
        checksum_addr = Web3.toChecksumAddress(account)
        root, is_whitelisted = self.contract.functions.getWhitelistedRoot(checksum_addr).call()
        last_auth = self.contract.functions.lastAuthenticated(checksum_addr).call()
        return {
            "is_whitelisted": is_whitelisted,
            "root": root,
            "last_authenticated_timestamp": last_auth,
        }

    async def get_expiry_data(self, account: str) -> dict:
        checksum_addr = Web3.toChecksumAddress(account)
        last_auth = self.contract.functions.lastAuthenticated(checksum_addr).call()
        auth_period = self.contract.functions.authenticationPeriod().call()
        expires_at = last_auth + (auth_period * 86400)
        import time
        is_expired = time.time() > expires_at
        return {
            "last_authenticated": last_auth,
            "authentication_period_days": auth_period,
            "expires_at": expires_at,
            "is_expired": is_expired,
        }


_identity_service = None


def get_identity_service() -> IdentityService:
    global _identity_service
    if _identity_service is None:
        _identity_service = IdentityService()
    return _identity_service
